doc/pekwm_bg.1
.nh
.TH pekwm\_bg 1 "February 2021" pekwm\_bg "User Manual"

.SH NAME
.PP
pekwm\_bg \- a simple background setting application k


.SH SYNOPSIS
.PP
pekwm\_bg [\-\-display] texture


.SH DESCRIPTION
.PP
pekwm\_bg is a simple background setting application bundled together
with pekwm and is typically run whenever a theme has a background
texture specified.

.PP
pekwm\_bg repeats the texture on each head, making images appear once
for every available head.

.PP
pekwm\_bg accepts textures in the same format as pekwm themes where the
most common ones for backgrounds are:

.RS
.IP \(bu 2
\fBImage\fP background.png#scaled, background.png scaled to screen size.
.IP \(bu 2
\fBSolid\fP #eeeeee, solid color.
.IP \(bu 2
\fBLinesHorz\fP 33% #afadbf #9f9daf #afadbf, 3 horizontal lines.

.RE


.SH OPTIONS
.PP
\fB\-\-daemon\fP, run as daemon.

.PP
\fB\-\-display\fP \fIDISPLAY\fP Connect to DISPLAY instead of DISPLAY set in environment.

.PP
\fB\-\-help\fP Show help information.

.PP
\fB\-\-load\-dir\fP DIR Load images from specified directory.

.PP
\fB\-\-stop\fP Stop running pekwm\_bg daemon.


NEWS
NEWS pekwm-0.1.7
================

Bugs fixed

  * #XX Several memory leaks identified with valgrind plugged.
  * #XX Fixed broken iconfied state reported by
     <giuliani{.}blackfenix{@}gmail{.}com>
  * #XX CfgParser parsed data twice.
  * #XX StatusWindow placement fixed.
  * #XX Fix for _NET_WM_NAME ops by Florian Hoppe <flp{@}floppi{.}de>
  * #22 Font type is now possible to define as the first argument in
     themes.
  * #30 Panel raise issues, reported and fixed by Jyri Jokinen
     <shared{@}adresh{.}com>
  * #21 Dynamic menu entries become unselectable after first
     visit. Reported by Lassi Pölönen and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #14 SendKey action requested by nidan.
  * #29 drag'n'drop requires the recieving window to be moved before
     working. Reported by Jyri Jokinen <shared{@}adresh{.}com>, fixed
     by Andreas <ioerror{@}lavabit{.}com>
  * #32 Shaping issues after shading. Reported by Paulo Igor Barra
     Nascimento <pibarnas{@}yahoo{.}com{.}br> and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #XX Corrects the interpretation of the motif windowmanager hints
     by Andreas <ioerror{@}lavabit{.}com>
  * #XX Remove perl scripts reducing dependencies on packages. 

User visible changes

  * UTF-8 support added for reading and rendering window titles.
  * _NET_WM_ICON is now read and rendered in menus.
  * RandR support for multi-screen configurations.
  * New default theme.
  * CmdDialog size is now calculated from the current head and not the
    whole display.

Configuration changes

  * _NET_WM_WINDOW_TYPE handling configurable via autoproperties.
  * Width/Height 0 in autoproperties now means screen width/height.
  * WorkspacesPerRow and WorkspaceNames added with matching
    GotoWorkspace/SendToWorkspace parameters Up, PrevV, Down, NextV.
  * ShowWorkspaceIndicator
  * WorkspaceIndicatorScale
  * PCRE support remove.
  * EdgeIndent option added to Screen section of main configuration
    file. If true reserves space where the screen edge is.
  * EdgeSize now supports per edge configuration. 

Theme changes

  * Shaping suport for non-corner windows.
  * XMB type is used by default in favour of X11 type fonts.
  * Font alpha is now handles as % from 0-100 instead of previous Xft
    unsigned int value.
  * HeightAdapt option added to theme.
  * Font type is now possible to define as the first argument in themes.


src/pekwm_bg.cc
//
// pekwm_bg.cc for pekwm
// Copyright (C) 2021 Claes Nästén <pekdon@gmail.com>
//
// This program is licensed under the GNU GPL.
// See the LICENSE file for more information.
//

#include "pekwm.hh"

#include "ImageHandler.hh"
#include "TextureHandler.hh"
#include "Util.hh"
#include "x11.hh"

extern "C" {
#include <getopt.h>
#include <signal.h>
#include <unistd.h>
}

static bool _stop = false;
static ImageHandler* _image_handler = nullptr;
static TextureHandler* _texture_handler = nullptr;

namespace pekwm
{
    ImageHandler* imageHandler()
    {
        return _image_handler;
    }

    TextureHandler* textureHandler()
    {
        return _texture_handler;
    }
}

static void sighandler(int signal)
{
    if (signal == SIGINT) {
        _stop = true;
    }
}

static void init(Display* dpy)
{
    _image_handler = new ImageHandler();
    _texture_handler = new TextureHandler();
}

static void cleanup()
{
    delete _texture_handler;
    delete _image_handler;
}

static void usage(const char* name, int ret)
{
    std::cout << "usage: " << name << " [-hl] texture" << std::endl
              << "  -d --display dpy    Display" << std::endl
              << "  -D --daemon         Run in the background" << std::endl
              << "  -h --help           Display this information" << std::endl
              << "  -l --load-dir path  Search path for images" << std::endl
              << "  -s --stop           Stop running pekwm_bg" << std::endl;
    exit(ret);
}

static Pixmap setBackground(PTexture *tex)
{
    auto pix = X11::createPixmap(X11::getWidth(), X11::getHeight());

    // render background per-head, might not suite all but leave it at
    // that for now.
    for (int i = 0; i < X11::getNumHeads(); i++) {
        auto head = X11::getHeadGeometry(i);
        tex->render(pix, head.x, head.y, head.width, head.height);
    }

    X11::setLong(X11::getRoot(), XROOTPMAP_ID, pix, XA_PIXMAP);
    X11::setLong(X11::getRoot(), XSETROOT_ID, pix, XA_PIXMAP);
    X11::setWindowBackgroundPixmap(X11::getRoot(), pix);
    X11::clearWindow(X11::getRoot());

    return pix;
}

void modeBackground(const std::string& tex_str)
{
    auto tex = pekwm::textureHandler()->getTexture(tex_str);
    if (tex) {
        std::cout << "Setting background " << tex_str << std::endl;
        auto pix = setBackground(tex);
        pekwm::textureHandler()->returnTexture(tex);

        // used for stop actions
        X11::setLong(X11::getRoot(), PEKWM_BG_PID, getpid());

        XEvent ev;
        while (! _stop && X11::getNextEvent(ev)) {
            ;
        }

        X11::freePixmap(pix);
    } else {
        std::cerr << "Failed to load texture " << tex_str << std::endl;
    }
}

void modeStop()
{
    long pid;
    if (! X11::getLong(X11::getRoot(), PEKWM_BG_PID, pid)) {
        std::cerr << "Failed to get _PEKWM_BG_PID, unable to stop" << std::endl;
        return;
    }

    kill(pid, SIGINT);

    std::cerr << "Sent " << pid << " SIGINT" << std::endl;
}

int main(int argc, char* argv[])
{
    const char* display = NULL;
    bool do_daemon = false;
    bool stop = false;
    std::string load_dir("./");

    static struct option opts[] = {
        {"display", required_argument, NULL, 'd'},
        {"daemon", no_argument, NULL, 'D'},
        {"help", no_argument, NULL, 'h'},
        {"load-dir", required_argument, NULL, 'l'},
        {"stop", no_argument, NULL, 's'},
        {NULL, 0, NULL, 0}
    };

    int ch;
    while ((ch = getopt_long(argc, argv, "d:Dhl:s", opts, NULL)) != -1) {
        switch (ch) {
        case 'd':
            display = optarg;
            break;
        case 'D':
            do_daemon = true;
            break;
        case 'h':
            usage(argv[0], 0);
            break;
        case 'l':
            load_dir = optarg;
            if (load_dir.back() != '/') {
                load_dir += '/';
            }
            Util::expandFileName(load_dir);
            break;
        case 's':
            stop = true;
            break;
        default:
            usage(argv[0], 1);
            break;
        }
    }

    std::string tex_str;
    for (int i = optind; i < argc; i++) {
        if (i > optind) {
            tex_str += " ";
        }
        tex_str += argv[i];
    }

    if (! stop && tex_str.empty()) {
        usage(argv[0], 1);
        return 1;
    }

    auto dpy = XOpenDisplay(display);
    if (! dpy) {
        std::cerr << "Can not open display!" << std::endl
                  << "Your DISPLAY variable currently is set to: "
                  << Util::getEnv("DISPLAY") << std::endl;
        return 1;
    }

    // setup signal handler after connecting to the display
    signal(SIGINT, &sighandler);

    X11::init(dpy, true);
    init(dpy);

    std::cout << "Load dir " << load_dir << std::endl;
    _image_handler->path_push_back(load_dir);

    modeStop();
    if (! stop) {
        if (do_daemon) {
            if (daemon(0, 0) == -1) {
                std::cerr << "Failed to daemonize: " << strerror(errno)
                          << std::endl;
            }
        }
        modeBackground(tex_str);
    }

    cleanup();
    X11::destruct();

    return 0;
}
